File: src/main/java/twopointers/TrappingRainWater.java
package twopointers;

public class TrappingRainWater {

    public int sol(int[] height) {
        if (height.length == 0) {
            return 0;
        }
        int left = 0;
        int right = height.length - 1;
        int max = 0;
        int leftMaxHeight= 0;
        int rightMaxHeight= 0;

        while (left < right) {
            leftMaxHeight = Math.max(leftMaxHeight, height[left]);
            rightMaxHeight = Math.max(rightMaxHeight, height[right]);

            if (leftMaxHeight < rightMaxHeight) {
                max += leftMaxHeight - height[left];
                left++;
            } else {
                max += rightMaxHeight - height[right];
                right++;
            }
        }
        return max;
    }
}

File: src/main/java/twopointers/TrappingRainWater.py
from typing import List


class Solution:
    def trap(self, height: List[int]) -> int:
        area = 0
        left_index = 0
        right_index = len(height) - 1
        left_max_height = 0
        right_max_height = 0

        while left_index < right_index:
            left_max_height = max(left_max_height, height[left_index])
            right_max_height = max(right_max_height, height[right_index])
            if left_max_height < right_max_height:
                area += left_max_height - height[left_index]
                left_index += 1
            else:
                area += right_max_height - height[right_index]
                right_index -= 1
        return area
